libr/core/canal.c

```diff
@@ -2753,39 +2753,90 @@ R_API void r_core_recover_vars(RCore *core, RAnalFunction *fcn, bool argonly) {
 	return;
 }
 
-R_API RList* r_core_anal_graph_to(RCore *core, ut64 addr, int n) {
-	RAnalBlock *bb, *root = NULL, *dest = NULL;
-	RListIter *iter, *iter2;
-	RList *list2 = NULL, *list = NULL;
-	RAnalFunction *fcn;
+static int cmp_addr(const void *a, const void *b) {
+	const ut64 *_a, *_b;
+
+	_a = a;
+	_b = b;
+
+	return *_a - *_b;
+}
+
+static bool anal_path_exists(RCore *core, ut64 from, ut64 to, RList *bbs, int depth, RList *state) {
+	RAnalBlock *bb = r_anal_bb_from_offset (core->anal, from);
+	RListIter *iter = NULL;
+	RList *refs = NULL;
+	RAnalRef *refi;
+	RAnalFunction *fcn = NULL;
+	RAnalFunction *cur_fcn = NULL;
+	ut64 *frp = R_NEW0 (ut64);
+
+	if (depth < 1) {
+		eprintf ("going too deep\n");
+		return false;
+	}
+
+	if (!bb || !bbs) {
+		return false;
+	}
+
+	*frp = from;
+	r_list_append (state, (void *)frp);
+
+	ut64 tj = bb->jump;
+	ut64 fj = bb->addr + bb->size;
+
+	// try to find the target in the current function
+	if (r_anal_bb_is_in_offset (bb, to) ||
+		((!r_list_find (state, (void *)&(tj), cmp_addr) && anal_path_exists (core, bb->jump, to, bbs, depth - 1, state))) ||
+		((!r_list_find (state, (void *)&(fj), cmp_addr) && anal_path_exists (core, bb->addr + bb->size, to, bbs, depth - 1, state)))) {
+		r_list_prepend (bbs, bb);
+		return true;
+	}
 
+	// find our current function
 	r_list_foreach (core->anal->fcns, iter, fcn) {
-		if (!r_anal_fcn_is_in_offset (fcn, core->offset)) {
-			continue;
+		if (r_anal_fcn_in (fcn, from)) {
+			cur_fcn = fcn;
+			break;
 		}
-		r_list_foreach (fcn->bbs, iter2, bb) {
-			if (r_anal_bb_is_in_offset (bb, addr)) {
-				dest = bb;
-			}
-			if (r_anal_bb_is_in_offset (bb, core->offset)) {
-				root = bb;
-				r_list_append (list, list2);
+	}
+
+	// get call refs from current basic block and find a path from them
+	if (cur_fcn) {
+		refs = r_anal_fcn_get_refs (core->anal, cur_fcn);
```

Comment:
I think you are not considering the "backward" edge of a call. The "ending" basic block of a function should have a link to the basicblocks that have a call to it.

Reply:
right ! thanks

Reply:
correct me if I am wrong, but I think this is in fact not needed.
If you have a look at the code:
https://github.com/radare/radare2/blob/db2a8619761c76dcc9f6aa3bef9c5dde4e2e7315/libr/core/canal.c#L2791
you see that it is only once all blocks are analysed (when we reach a leaf block) that anal_path_exists will start following calls.

Reply:
I'm not sure I follow you... what I'm saying is: suppose you have function A that calls function B. Your seek is at the last basic block of function B. Now you do `abt 0xYYYYYYY` where 0xYYYYYYY is an address in the function A, a few basic blocks after the call to function B. I would expect `abt` to give me the path that includes the `ret` instruction in the last basic block of function B. Not sure if it's clearer :)

Reply:
pretty clear !

Reply:
fixed